R/site.R

```diff
@@ -0,0 +1,191 @@
+#'
+#' \itemize{
+#'   \item{The active output format defines a "_site" function (e.g.
+#'   \code{html_document_site}).}
+#'   \item{The \code{site} metadata entry refers to a custom function (e.g.
+#'   \code{mypackage::site}).}
+#' }
+#'
+#' The site generation function should return a list with the following
+#' elements:
+#'
+#' \itemize{
+#'   \item{\code{name}} {The name for the website (e.g. the parent directory name).}
+#'   \item{\code{output_dir} {The directory where the website output is written to
+#'   (e.g. "." or "_site")}}
+#'   \item{\code{render}} {An R function that can be called to generate the site.
+#'   The function should accept the \code{output_format}, \code{quiet},
+#'   \code{encoding}, and \code{...} arguments}
+#' }
+#'
+#' See the source code of the \code{rmarkdown::default_site} function for a
+#' simple example of a custom site generation function.
+#'
+#' @section Site Preview:
+#'
+#' Note that the \code{render_site} function can be used as a custom \code{knit}
+#' handler within RStudio by adding the following to the YAML of any
+#' document within a site directory:
+#'
+#' \code{knit: rmarkdown::render_site}
+#'
+#' This will result in the entire site being rendered each time Knit is
+#' executed within RStudio.
+#'
+#' @export
+render_site <- function(input = ".",
+                        output_format = NULL,
+                        quiet = FALSE,
+                        encoding = getOption("encoding"),
+                        ...) {
+
+  # normalize input (capture original input first)
+  original_input <- input
+  input <- input_as_dir(input)
+
+  # find the site generator
+  generator <- site_generator(input, output_format, encoding)
+  if (is.null(generator))
+    stop("No index.Rmd with site entry found.")
+
+  # execute it
+  generator$render(output_format = output_format,
+                   quiet = quiet,
+                   encoding = encoding)
+
+  # print the name of the index file (for RStudio Preview)
+  if (!quiet) {
+    # if the input was a filename use that as a base (otherwise
+    # use index.Rmd)
+    if (!dir_exists(original_input))
+      output <- file_with_ext(basename(original_input), "html")
+    else
+      output <- "index.html"
+    output <- file.path(input, generator$output_dir, output)
+    message("\nOutput created: ", relative_to(getwd(), output))
+  }
+}
+
+#' @rdname render_site
+#' @export
+site_generator <- function(input = ".",
+                           output_format = NULL,
+                           encoding = getOption("encoding")) {
+
+  # normalize input
+  input <- input_as_dir(input)
+
+  # if we have an index.Rmd then check it's yaml for "site:"
+  index <- file.path(input, "index.Rmd")
+  if (file.exists(index)) {
+
+    # read index.Rmd and extract the front matter
+    index_lines <- read_lines_utf8(index, encoding)
+    front_matter <- parse_yaml_front_matter(index_lines)
+    if (!is.null(front_matter) && !is.null(front_matter$site)) {
```

Comment:
Can `parse_yam_front_matter` ever return null? If yes, then we are missing similar null-checks at the other call locations.